### PR TITLE
[DataGrid] Keep focused cell in the DOM

### LINKED ISSUE #5750: [data grid] Keyboard navigation doesn't work when active cell is out of the viewport
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Whenever I press any of the arrow keys, it applies scroll to the window.
https://codesandbox.io/s/1cm3xe?file=/demo.tsx

https://user-images.githubusercontent.com/4550843/183926380-7b5ce2a4-d6e0-4d5f-bbaa-13314092a9b9.mov

### Expected behavior 🤔

DataGrid should select the next cell and scroll it into view.
The same way it works in [ag-grid](https://www.ag-grid.com/).

https://user-images.githubusercontent.com/4550843/183927567-056100ca-62f4-4b5b-8558-659279faddca.mov

### Steps to reproduce 🕹

_No response_

### Context 🔦

_No response_

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

### Order ID 💳 (optional)

_No response_

### LINKED ISSUE #6165: [data grid] `onRowOrderChange` is not called sometimes when dragging causes scrolling in datagrid.
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

When dragging a row up or down and it causes scrolling of the table, it sometimes does not call onRowOrderChange. Seems to happen most when scrolling a long way.
The row is moved correctly in the grid and no errors shown in console. Dragging it again a short distance does trigger onRowOrderChange.

### Expected behavior 🤔

onRowOrderChange should be called whenever a row is dragged.

### Steps to reproduce 🕹

Link to live example:

Steps:
1.
2.
3.
4.

### Context 🔦

_No response_

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

### Order ID 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by yaredtsy**
**Problem**

When `virtualization` is enabled, the focused `cells` will be unmounted when they get out of range. This causes a problem if you are `reordering` or navigating with the keyboard.

**inspiration** 

I looked at how AG-grid implemented their virtualization. and what I observed is they keep the focused `Row` always in the table.

https://user-images.githubusercontent.com/34584266/210151091-0f34ec09-be23-4f49-a5b1-5b07e914189f.mp4

**Solution**

the `row` which holds the focused `cell` should always be rendered, even if it's out of range it should always be in `renderedRows` 

**Fixes**

- [x] DataGrid reorder with virtualization, Fixes #6165
- [X] keyboard navigation for a cell that is out of viewport(Vertical virtualization), Fixes #5750 

**Todo**

- [x] keyboard navigation for a cell that is out of viewport( horizontal virtualization)
- [x] Close #6765

### CODE REVIEW

**Comment by yaredtsy** [comment]
@cherniavskii , @m4theushw

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx [hunk 1]
```diff
@@ -418,6 +418,9 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
     const rowBuffer = !disableVirtualization ? rootProps.rowBuffer : 0;
     const columnBuffer = !disableVirtualization ? rootProps.columnBuffer : 0;
 
+    let isFocusedCellRendered = false;
+    const { cell } = apiRef.current.state.focus;
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:422]:
```suggestion
    const focusedCell = gridFocusCellSelector(apiRef);
```
Use the selector instead

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx [hunk 2]
```diff
@@ -453,6 +460,29 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
...
+      // If the selected row is not within the current range of rows being displayed,
+      // we need to render it at either the top or bottom of the rows,
+      // depending on whether it is above or below the range.
+      if (!isFocusedCellRendered && cell && !disableVirtualization) {
+        const rows = currentPage.rows.filter((row) => row.id === cell.id);
+        const focusedRow = rows.length > 0 && rows[0];
+
+        if (focusedRow) {
+          const index = currentPage.rows.indexOf(focusedRow);
+          if (index > firstRowToRender) {
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:473]:
```suggestion
        const focusedRowIndex = currentPage.rows.findIndex((row) => row.id === cell.id);
        const focusedRow = currentPage.rows[focusedRowIndex];

        if (focusedRow) {
          if (focusedRowIndex > firstRowToRender) {
```

Use `.findIndex` instead of `.filter` + `.indexOf`?

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx [hunk 3]
```diff
@@ -453,6 +460,29 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
...
+            renderedRows.push(focusedRow);
+          } else {
+            renderedRows.unshift(focusedRow);
+          }
+          apiRef.current.calculateColSpan({
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:478]: It doesn't seem to be working with pinned columns (and column spanning?):
https://deploy-preview-7357--material-ui-x.netlify.app/x/react-data-grid/column-spanning/#function-signature


https://user-images.githubusercontent.com/13808724/212136172-443d0379-f8dd-4be6-b9d7-3ceb61834322.mov

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx [hunk 4]
```diff
@@ -453,6 +460,29 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
           columns: visibleColumns,
         });
       }
+
+      // If the selected row is not within the current range of rows being displayed,
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:464]: The `rowThreshold` doesn't work as expected. You can see the focused cell before virtualized rows are rendered: https://deploy-preview-7357--material-ui-x.netlify.app/x/react-data-grid/virtualization/

https://user-images.githubusercontent.com/13808724/212136668-2c878c74-67a3-4681-a9d3-abcd64dad4f8.mov

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx [hunk 5]
```diff
@@ -473,11 +503,34 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
 
     const renderedColumns = visibleColumns.slice(firstColumnToRender, lastColumnToRender);
 
+    // If the selected column is not within the current range of columns being displayed,
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:506]: There's a focus jump when clicking on the cell after the focused cell was moved out of view: https://deploy-preview-7357--material-ui-x.netlify.app/x/react-data-grid/virtualization/

https://user-images.githubusercontent.com/13808724/212136876-48905cf4-bc80-4e00-a41d-d63fcfd690e7.mov

**Review by cherniavskii — CHANGES_REQUESTED** [review]
Thanks, Yared!
I like the idea!
Here's my initial feedback on the proposed changes

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx (hunk 2 above)
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx]: done.I have also moved it out of the function so it won't have to find the index every time if the focused cell is not changed

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx (hunk 4 above)
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx]: I have set the row width and opacity to zero. it seems to fixe the problem

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx (hunk 5 above)
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:516]: The focused cell was causing the render zone to shift to the right, resulting in the appearance of a jumping focus. To resolve this, I have set the cell width and opacity to zero when using column virtualization.

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx [hunk 6]
```diff
@@ -418,6 +418,9 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
     const rowBuffer = !disableVirtualization ? rootProps.rowBuffer : 0;
     const columnBuffer = !disableVirtualization ? rootProps.columnBuffer : 0;
 
+    let isFocusedCellRendered = false;
```
[the remark this answers is not in the document]
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx]: I think its a good idea . I refactored most of the code. I inverted the flag and it simplify the conditions. I rename it to `isRowWithFocusedCellNotInRange`.  and the only way it going to be true is if there is a focused cell and it's not in range.
```jsx
    let isRowWithFocusedCellNotInRange = false;

    if (indexOfRowWithFocusedCell > -1) {
      const rowWithFocusedCell = currentPage.rows[indexOfRowWithFocusedCell];
      if (
        firstRowToRender > indexOfRowWithFocusedCell ||
        lastRowToRender < indexOfRowWithFocusedCell
      ) {
        isRowWithFocusedCellNotInRange = true;
        if (indexOfRowWithFocusedCell > firstRowToRender) {
          renderedRows.push(rowWithFocusedCell);
        } else {
          renderedRows.unshift(rowWithFocusedCell);
        }
        apiRef.current.calculateColSpan({
          rowId: rowWithFocusedCell.id,
          minFirstColumn,
          maxLastColumn,
          columns: visibleColumns,
        });
      }
    }
```
i have changed to this so  there will consistency with column virtualization

File: packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx (hunk 4 above)
Comment (cherniavskii) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx]: Yes, I cannot see the row with the focused cell, but now I see the blank space, which shouldn't happen either:

https://user-images.githubusercontent.com/13808724/219099182-6e8c1ea0-03ff-4662-be83-5eabfbbb3076.mov

Compared to https://deploy-preview-7357--material-ui-x.netlify.app/x/react-data-grid/virtualization/ where it doesn't happen:

https://user-images.githubusercontent.com/13808724/219099683-eaea2a3b-e5bf-4b57-b045-36a9a776591d.mov

Because the focused row is rendered when it's out of the viewport, it shifts the rows below it and therefore breaks the rowBuffer behavior

Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx]: I forgot to set the row height to 0. its fixed now

Comment (yaredtsy) [reply packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx]: I think it's resolved now.

**Comment by gidztech** [comment]
I think this **might** fix a bug I'm having where clicking a button within a row column sometimes makes the scrollbar jump when virtualization is enabled, though I'm not 100%.

**Comment by cherniavskii** [comment]
@m4theushw Could you review this PR? It would be great to merge it soon

**Comment by yaredtsy** [comment]
It is now compatible with the memoization feature.

**Comment by romgrk** [comment]
@yaredtsy if you're still interested in merging this, can you merge/rebase master? I'll review as soon as it's done so we can get this in.

File: packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx [hunk 7]
```diff
@@ -81,6 +81,7 @@ export const useGridRowReorder = (
       });
 
       originRowIndex.current = apiRef.current.getRowIndexRelativeToVisibleRows(params.id);
+      apiRef.current.setCellFocus(params.id, '__reorder__');
```
Comment (romgrk) [inline packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx:84]: Do we have a constant for this? I like to avoid magic strings if possible.

Also, why is this change necessary?

File: packages/grid/x-data-grid/src/components/GridRow.tsx [hunk 8]
```diff
@@ -57,6 +57,8 @@ export interface GridRowProps extends React.HTMLAttributes<HTMLDivElement> {
   tabbableCell: string | null;
   row?: GridRowModel;
   isLastVisible?: boolean;
+  indexOfColumnWithFocusedCellNotInRange?: number;
+  isNotVisible?: boolean;
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/components/GridRow.tsx:61]: Is there a way we can reduce the name of `indexOfColumnWithFocusedCellNotInRange` a bit?

File: packages/grid/x-data-grid/src/components/GridRow.tsx [hunk 9]
```diff
@@ -277,6 +281,16 @@ const GridRow = React.forwardRef<HTMLDivElement, GridRowProps>(function GridRow(
         Object.keys(editRowsState).length > 0);
 
     const editCellState = editRowsState[rowId]?.[column.field] ?? null;
+    let cellIsNotVisible = false;
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/components/GridRow.tsx:284]: I haven't tested live yet, but could you summarize briefly what rows/columns we render when the focused cell is outside the range?

File: packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx (hunk 7 above)
Comment (yaredtsy) [reply packages/grid/x-data-grid-pro/src/hooks/features/rowReorder/useGridRowReorder.tsx]: > Also, why is this change necessary?

It's fixes https://github.com/mui/mui-x/issues/6165. when you drag rows, you are really dragging the first draggable cell and it will not update the `focusedCell` state when you drag. so if you manage the cell to be out of the viewport and the cell is unmounted. it prevents the rows from reordering.

File: packages/grid/x-data-grid/src/components/GridRow.tsx (hunk 8 above)
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/components/GridRow.tsx:61]: what about this `focusedCellColumnIndexNotInRange`?

File: packages/grid/x-data-grid/src/components/GridRow.tsx [hunk 10]
```diff
@@ -277,6 +281,16 @@ const GridRow = React.forwardRef<HTMLDivElement, GridRowProps>(function GridRow(
...
+
+    if (
+      indexOfColumnWithFocusedCellNotInRange &&
+      visibleColumns[indexOfColumnWithFocusedCellNotInRange].field === column.field
+    ) {
```
[the remark this answers is not in the document]
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/components/GridRow.tsx]: yep it did, fixed

File: packages/grid/x-data-grid/src/components/GridRow.tsx (hunk 8 above)
Comment (romgrk) [reply packages/grid/x-data-grid/src/components/GridRow.tsx]: Is the `NotInRange` strictly needed here?

Comment (yaredtsy) [reply packages/grid/x-data-grid/src/components/GridRow.tsx]: yes. If the focused cell is in range, it will be undefined. We need to find a way to indicate that the index is only held when it falls out of range.

File: packages/grid/x-data-grid/src/components/GridRow.tsx [hunk 11]
```diff
@@ -370,7 +396,18 @@ const GridRow = React.forwardRef<HTMLDivElement, GridRowProps>(function GridRow(
...
+
+    if (
+      focusedCellColumnIndexNotInRange &&
+      visibleColumns[focusedCellColumnIndexNotInRange].field === column.field &&
+      focusedCell
+    ) {
+      indexRelativeToAllColumns = focusedCellColumnIndexNotInRange;
+    } else if (focusedCellColumnIndexNotInRange && focusedCell) {
+      indexRelativeToAllColumns -= 1;
+    }
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/components/GridRow.tsx:410]: The condition checks in this section fail for `0` as well I think.

File: packages/grid/x-data-grid/src/components/cell/GridCell.tsx [hunk 12]
```diff
@@ -45,7 +44,7 @@ type GridCellV7Props = {
   width: number;
   colSpan?: number;
   disableDragEvents?: boolean;
-  editCellState: GridEditCellProps<any> | null;
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/components/cell/GridCell.tsx:48]: Why is this one removed? Was it unused?

**Review by romgrk — COMMENTED** [review]
You need to run `yarn proptypes` after doing modifications to the typescript props to pass the CI.

File: packages/grid/x-data-grid/src/components/GridRow.tsx [hunk 13]
```diff
@@ -277,6 +281,16 @@ const GridRow = React.forwardRef<HTMLDivElement, GridRowProps>(function GridRow(
...
+    let cellIsNotVisible = false;
+
+    if (
+      focusedCellColumnIndexNotInRange !== undefined &&
+      visibleColumns[focusedCellColumnIndexNotInRange].field === column.field
+    ) {
+      cellIsNotVisible = true;
+    } else {
+      cellIsNotVisible = false;
+    }
```
[the remark this answers is not in the document]
Comment (yaredtsy) [reply packages/grid/x-data-grid/src/components/GridRow.tsx]: removed

File: packages/grid/x-data-grid/src/components/GridRow.tsx (hunk 11 above)
Comment (romgrk) [reply packages/grid/x-data-grid/src/components/GridRow.tsx]: One of them is fixed but the other remains, there are 2 checks in this block.

**Comment by yaredtsy** [comment]
@romgrk I am having trouble with `yarn proptypes` it's not generating as the pipeline expected. mine is not adding `.required`. on required fields. is there something I must do?.

**Comment by romgrk** [comment]
I'm not sure what's going on, the `.isRequired` flag should not be changed by this PR. Maybe the build script is confused. I'd try to play with it to see if something in the changes causes the confusion. As a last resort, fixing the propTypes manually would be fine.

**Comment by yaredtsy** [comment]
@romgrk its fixed now

File: packages/grid/x-data-grid/src/components/GridRow.tsx [hunk 14]
```diff
@@ -322,27 +335,38 @@ const GridRow = React.forwardRef<HTMLDivElement, GridRowProps>(function GridRow(
...
+      maxHeight: rowHeight === 'auto' ? 'none' : rowHeight, // max-height doesn't support "auto"
+      minHeight,
+    };
+    if (isNotVisible) {
+      return {
+        opacity: 0,
+        width: 0,
+        height: 0,
+      };
+    }
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/components/GridRow.tsx:350]: Move the `if/return` to the start of this function. Apply the same pattern in `GridCell` where appropriate.

**Review by romgrk — COMMENTED** [review]
The feature is working when the focused cell is above the rendered region, but not when it's below. To test it I've put a cell in edit mode, scrolled up, and at some point the cell goes away.

Also I'm not sure what behavior we expect/want/accept when a user does a keyboard-input while the focused cell is outside of range, but currently it feels a bit off, the browser "brings back" the element in view but it's `width: 0; height: 0` so it just ends up scrolling to the top of the current render range instead of to where the cell is. Not sure what's the best behavior here.

**Comment by yaredtsy** [comment]
> The feature is working when the focused cell is above the rendered region, but not when it's below. To test it I've put a cell in edit mode, scrolled up, and at some point the cell goes away. I can provide a video if you can't reproduce it.

yes when it is in edit mode it wont scroll b/c `scrollToIndexes`, wont be called, also if you hit `Tab` it will not scroll. 

> Also I'm not sure what behavior we expect/want/accept when a user does a keyboard-input while the focused cell is outside of range

i have tried to adpat what ag-grid is doing.

**Comment by yaredtsy** [comment]
> The feature is working when the focused cell is above the rendered region, but not when it's below. To test it I've put a cell in edit mode, scrolled up, and at some point the cell goes away. I can provide a video if you can't reproduce it.

yes when it is in edit mode it wont scroll b/c `scrollToIndexes`, wont be called, also if you hit `Tab` it will not scroll. 

> Also I'm not sure what behavior we expect/want/accept when a user does a keyboard-input while the focused cell is outside of range

i have tried to adpat what ag-grid is doing.

**Comment by romgrk** [comment]
Alright, I think this behavior is acceptable.

Could you address the comment I left above? Could you also add tests to cover this feature? Once that is done, this should be ready.

**Comment by yaredtsy** [comment]
@romgrk done

File: packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx [hunk 15]
```diff
@@ -184,6 +185,42 @@ describe('<DataGrid /> - Cells', () => {
...
+    const thirdRowCell = getCell(1, 3);
+    userEvent.mousePress(thirdRowCell);
+
+    const activeElementTextContent = document.activeElement?.textContent;
+    const columnWidth = document.activeElement!.clientWidth;
+
+    // check for row virtualazition
+    const scrollTop = 10 * rowHeight;
+    virtualScroller.scrollTop = scrollTop;
+    virtualScroller.dispatchEvent(new Event('scroll'));
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx:212]: Any way to use `fireEvent` like the rest of the test code does?

https://stackoverflow.com/a/56233748/6303229

Also applies below

File: packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx [hunk 16]
```diff
@@ -184,6 +185,42 @@ describe('<DataGrid /> - Cells', () => {
     }).toWarnDev(['MUI: The cell with id=1 and field=brand received focus.']);
   });
 
+  it('selected cell  should alwys been in the dom,', function test() {
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx:188]: Typos & orthograph errors. Can you also start it so it reads naturally like the other labels? E.g. `it('should keep selected ...')`

File: packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx [hunk 17]
```diff
@@ -184,6 +185,42 @@ describe('<DataGrid /> - Cells', () => {
...
+
+    const virtualScroller = document.querySelector('.MuiDataGrid-virtualScroller')!;
+
+    const thirdRowCell = getCell(1, 3);
+    userEvent.mousePress(thirdRowCell);
+
+    const activeElementTextContent = document.activeElement?.textContent;
+    const columnWidth = document.activeElement!.clientWidth;
+
+    // check for row virtualazition
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx:209]: These comments aren't really necessary, if the `it(...)` label is clear.

File: packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx [hunk 18]
```diff
@@ -184,6 +185,37 @@ describe('<DataGrid /> - Cells', () => {
     }).toWarnDev(['MUI: The cell with id=1 and field=brand received focus.']);
   });
 
+  it('Should be in the DOM even if when the focused cell is out of viewport.', function test() {
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx:188]: Start with a lowercase, like the other test labels, no final period. The sentence could also be reworked to be a bit more clear.

**Comment by romgrk** [comment]
You need to apply `yarn prettier`.

File: packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx [hunk 19]
```diff
@@ -184,6 +185,37 @@ describe('<DataGrid /> - Cells', () => {
     }).toWarnDev(['MUI: The cell with id=1 and field=brand received focus.']);
   });
 
+  it('should not change the active component when the active focused cell gets out of viewport', function test() {
```
Comment (romgrk) [inline packages/grid/x-data-grid/src/tests/cells.DataGrid.test.tsx:188]: `it('should keep the focused cell/row rendered in the DOM if it scrolls outside the viewport')`

**Review by romgrk — APPROVED** [review]
@mui/xgrid If someone else wants to add something here please do so, afaict this is ready for merging.

**Comment by cherniavskii** [comment]
@yaredtsy Thanks for working on this! 🎉
